### PR TITLE
snapshot: don't schedule next snapshot job for a removed volume

### PR DESCRIPTION
**Pull request by rohityadavcloud**
When management server starts, it starts the snapshot scheduler. In case there is a volume snapshot policy which exists for a volume which does not exist, it can cause SQL constraint issue and cause the management server to break from starting its various components and cause HTTP 503 error.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java [hunk 1]
```diff
@@ -378,6 +378,12 @@ public Date scheduleNextSnapshotJob(final SnapshotPolicyVO policy) {
         if (policyId == Snapshot.MANUAL_POLICY_ID) {
             return null;
         }
+
+        if (_volsDao.findById(policy.getVolumeId()) == null) {
+            s_logger.warn("Found snapshot policy ID: " + policyId + " for volume ID: " + policy.getVolumeId() + " that does not exist or has been removed");
+            return null;
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java:385]: great find, should we also delete the snapshotpolicy while we are at it?

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by weizhouapache** [comment]
what if the volume is removed manually after snapshot policy is created ?

it may be better to add constraint to the table `snapshot_policy`, to prevent the volume to be removed manually, if there are snapshot policies linked to it.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java (hunk 1 above)
Comment (rohityadavcloud) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java:385]: Makes sense

**Comment by DaanHoogland** [comment]
> what if the volume is removed manually after snapshot policy is created ?
> 
> it may be better to add constraint to the table `snapshot_policy`, to prevent the volume to be removed manually, if there are snapshot policies linked to it.

I agree, but this code is still good and anybody could use it in older versions (backport it) I don't think it hurts in any way even with the constraint.

If that works a cascading delete would be usefull in this case ;)

**Comment by weizhouapache** [comment]
> > what if the volume is removed manually after snapshot policy is created ?
> > it may be better to add constraint to the table `snapshot_policy`, to prevent the volume to be removed manually, if there are snapshot policies linked to it.
> 
> I agree, but this code is still good and anybody could use it in older versions (backport it) I don't think it hurts in any way even with the constraint.
> 
> If that works a cascading delete would be usefull in this case ;)

@DaanHoogland 
yes, with PR, the mgmt server will be started without any issue , even if volume is removed manually from database.

@rohityadavcloud 
code lgtm

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by rajujith** [comment]
> what if the volume is removed manually after snapshot policy is created ?
> 
> it may be better to add constraint to the table `snapshot_policy`, to prevent the volume to be removed manually, if there are snapshot policies linked to it.

Makes sense to add it.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java (hunk 1 above)
Comment (rohityadavcloud) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotSchedulerImpl.java]: Added code to remove schedule when volume is missing

**Comment by rohityadavcloud** [comment]
@weizhouapache I like your idea, could you propose that as a separate PR.

For now I've added changes to remove the schedule when volume is missing. Requesting re-review cc @weizhouapache @DaanHoogland @sureshanaparti @JoaoJandre 

@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by JoaoJandre — APPROVED** [review]
CLGTM

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by rohityadavcloud** [comment]
@JoaoJandre since you're the RM who has called off a freeze, could you merge this?

**Comment by rohityadavcloud** [comment]
This is an edge case not easy to reproduce but found in a prod customer env @JoaoJandre. To reproduce you can delete a volume from the db manually which has a snapshot policy.

**Comment by JoaoJandre** [comment]
> This is an edge case not easy to reproduce but found in a prod customer env @JoaoJandre. To reproduce you can delete a volume from the db manually which has a snapshot policy.

@rohityadavcloud  I was able to test it by deleting the volume on the DB while the MGMT was down. This really is an edge case, but the PR does not introduce any regressions and is adding a layer of safety to the snapshot policy process, so I don't see any issues in merging.
